packages/devextreme/js/__internal/ui/collection/m_collection_widget.edit.ts

```diff
@@ -428,11 +446,31 @@ const CollectionWidget = BaseCollectionWidget.inherit({
     this._setAriaSelectionAttribute($itemElement, String(isSelected));
   },
 
+  onSelectionChangingHandler(args) {
+    this._actions.onSelectionChanging(args);
+  },
+
+  _processSelectionChanging(args) {
+    this.onSelectionChangingHandler(this.selectionChangingArgs);
+    if (isPromise(this.selectionChangingArgs.cancel)) {
+      this.selectionChangingArgs.cancel.then((cancel) => {
+        if (!cancel) {
+          this.option('selectedItems', this._getItemsByKeys(args.selectedItemKeys, args.selectedItems));
+          this._updateSelectedItems(args);
+        }
+      }).catch((e) => {
+        console.error(e);
+      });
+    } else if (!this.selectionChangingArgs.cancel) {
+      this.option('selectedItems', this._getItemsByKeys(args.selectedItemKeys, args.selectedItems));
+      this._updateSelectedItems(args);
+    }
```

Comment:
[copypaste]
Can we reduce copypaste? You run almost the same code when cancel is promise and when is not